Code Review of TFMini in develop

jee7s opened the pull request.

jee7s committed:
Adding basic 1D KF for position estimation

jee7s committed:
correcting typos

jee7s committed:
Added initialize method and javadoc comments

jee7s committed:
Merge branch 'kf_dev' into develop

hoo761 committed:
Added a handler class for the TFMini LiDAR sensor **NOT TESTED**

hoo761 committed:
Fixed classpath entry in .classpath

hoo761 committed:
Fixed LiDAR code

jee7s reviewed (commented):
There appears to be an error in the state machine when checking the second byte

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -56,31 +56,41 @@
     }
     
+    private enum ReceivingState
+    {
```
These are the correct states for the state machine.

RG8452 replied:
The intent of the enum and reading method that uses the switch is to read each byte one at a time and whenever an actual packet is found, then parse through it and store it in the byte[] packet. I'm not sure if there's any reason not to lengthen the buffer, but I do know that Java does not have unsigned data types at all so we can't use a uint.

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -204,25 +170,65 @@ private void runLiDAR()
+                                   state = ReceivingState.WAIT_START2;
+                               }
+                               break;
+                           case WAIT_START2:
```
This WAIT_START2 state appears to be inspecting the same location in the packet as WAIT_START1. Or is the intent that we only fetch one byte at a time?

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -204,25 +170,65 @@ private void runLiDAR()
+                                   {
+                                       parsePacket();
+                                   }
+                                   state = ReceivingState.WAIT_START1;
```
Doesn't the checksum need to be reset to 0 somewhere in here?

jee7s replied:
Nevermind, we reset it on the first valid byte.

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -204,25 +170,65 @@ private void runLiDAR()
     
+    public void parsePacket()
+    {
+        distance = (((int) packet[3] * 256) + ((int) packet[2]));
```
I think these should be uints, else 0xFF = -128 rather than the 256 we want it to be.

dawilson9 replied:
You are correct (except for the 0xFF = -128 thing coz it actually equals -1 if we're using 8 bit signed integers or 255 for unsigned.) My assumption, though, is that "byte" is unsigned so the 0xFF we read there will be cast to 255 as an integer (signed or unsigned). Maybe Java works differently from C, of course?

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -171,7 +145,8 @@ public double getSignalQualityDegree()
     {
     	sensor.write(config, config.length);
-    	sensor.setTimeout(1/100);				// Sensor is 100Hz, so timeout is .001 seconds
+    	sensor.setReadBufferSize(PACKET_LENGTH);
```
Might we want to make the read buffer longer than the packet? If we make it the length of the packet and fall behind in processing, we'll lose data.

dawilson9 replied:
We don't want old packets to be read so having the buffer as small as possible seemed like a sensible thing to do. at 115kbps with a 9 byte packet and 100 packets per second, 99% of the time, the line is idle so I expect we should be safe to buffer no more than 1 packet. It's probably OK to double this just in case, though.

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -204,25 +170,65 @@ private void runLiDAR()
+                                   // Reached the end of the packet, check checksum
+                                   if(checksum == bytes[0])
+                                   {
+                                       parsePacket();
```
For now, we may want to calculate the checksum but not prevent packet parse if it is bad, just to see if we are receiving some data.

dawilson9 replied:
We save every packet in the packet[] member regardless of whether it's good or bad. This could be public and dumped to the dashboard, though, for debug purposes.

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -204,25 +170,65 @@ private void runLiDAR()
+        distance = (((int) packet[3] * 256) + ((int) packet[2]));
+        strength = (((int) packet[5] * 256) + ((int) packet[4]));
+        quality = (int)packet[7];
+    }
```
Also, don't we need a synchronized(this) in here to make sure we're OK to write the new values and aren't switched out in the middle of these three writes?

dawilson9 replied:
Probably a good idea - this should be a critical section (by whatever means you use to do this in Java). I'm not worried about new packets being read as we're processing old one since this can't happen given the way the state machine is structured but someone querying distance, strength or quality could get values from different packets in some cases (though this is unlikely to cause problems either, I suspect).

jee7s commented on SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java:
```diff
@@ -183,15 +158,6 @@ private int amountOfBytes()
-    
     /**
      * Runs a loop in a Thread that will read the 9 byte packets and set each
      * byte to the correct piece of data for use if the data is valid.
```
Do we want this to be a thread within the class, or should the class implement Thread? I can see some problems arising if we call start() or runLiDAR() more than once, as each call would spawn a new thread on the same object.

dawilson9 replied:
It should definitely be a single instance thread but I don't know how to achieve this in Java. Each object instantiated should use the distance, strength and quality value read by the single thread. That also means that the packet member must be global too. I seem to remember there's some modifier you can add to the member definition that makes this occur. Any idea what it is?

Commit:
Moved EJML JARs around and fixed .classpath to make the KF code link.

Commit:
Changed INCHESPERSECOND to MAXINCHESPERSECOND

Just being persnickety.

Commit:
Tidied up TFMini code and move to MXP serial port

Removed redundant code.
Added commenting in the state machine.
Removed tabs from source.

Commit:
Switch from no parity to space parity for TFMini

The datasheet for the TFMini LIDAR lists the serial parity setting as "0" which I read as meaning no parity but, given other hints elsewhere in the datasheet, I now think this is actually space parity which pulls the parity bit of the character low in all cases. I've modified the code to use this setting instead if kNone.

jee7s committed:
Changed TFMini to be a singleton that extends thread.

hoo761 committed:
Changed parity to kNone to match tests, also fixed braces

hoo761 committed:
UIL State Code

Code at the end of State UIL

hoo761 committed:
Cleaned up warnings

hoo761 committed:
Removed .classpath, updated gitignore
